### ISSUE TITLE
CI/AppVeyor is too slow! 

### ISSUE THREAD

**Issue by bendichter** [NeurodataWithoutBorders/pynwb#712]
AppVeyor is causing us some headaches. It's way slower than the other CI test and it's holding back development. Would it be possible to speed it up or reduce the number down to a single test?

**Comment by bendichter** [comment]
@ajtritt

**Comment by dorukozturk** [comment]
@bendichter I think it is good to support different versions of python on windows,
assuming there are windows users of pynwb. When you say it is holding back development what do you mean?

In theory it should only delay merging a PR, not your development, unless you are relying on CI to run tests while developing. Is that the case?

I think a faster feedback loop for development would be to develop locally, use tox locally on your OS, get all tests passing locally via tox, then push a commit up to have CI tests run on other platforms before merging.

**Comment by bendichter** [comment]
What I mean is that it takes along time to merge more than one PR. We generally have a queue of PRs and when we merge one, the CI needs to start over for the next one. This isn't a problem for CircleCI because it's fast, but AppVeyor prevents us from going down the row and merging all of the PRs that are ready. We often don't get very far down the list in a single session and by the time we are at it again, the list has grown. This slows down development because sometimes PR merges cause merging conflicts down the line, but we don't know that until we merge, which is delayed, so we end up spending a lot of time maintaining PRs. If there is any way to speed up these tests, it would help us a lot.

I'm also having an issue when I run `tox` locally that is says I don't have `pandas` installed, which I don't really understand, but that's a separate issue.

**Comment by t-b** [comment]
Travis has nowadays also Windows support. Maybe they are faster. See https://blog.travis-ci.com/2018-10-11-windows-early-release.

**Comment by dorukozturk** [comment]
Good news for travis community. Though travis is slower than circleci so my guess would be slower windows builds just like Appveyor. 

@bendichter For the macOS tests, we run them when the PR is merged only. So we always check the status of master as opposed to individual PRs. Do you feel comfortable doing the same for windows? If yes I can start investigating that and hopefully that could be a way to move forward.

On a clean virtual environment you should be able to 
```
cd pynwb
pip install tox
tox
```

Let me know if that does not work for you.

**Comment by bendichter** [comment]
How would you feel about a compromise where 1 windows test is run before
merge and the rest are run after?

**Comment by dorukozturk** [comment]
I don't think appveyor gives us that much freedom when it comes to adding logic.
I think embedding too much logic in CI configuration files is an anti-pattern and will be hard to maintain
if something breaks on the CI end.

If we run the tests only after a merge and if something is broken the badge will go red.
So I think it will be noticed and I feel safe about that approach.

The tradeoff here is the psychological destruction after merging a PR when you think everything is working and now all of a sudden something is broken on master and you have to work on it again. Kind of a POST MERGE SYNDROME (this is my humble opinion). Though this never happened with macOS tests as far as I know.

**Comment by ajtritt** [comment]
I think the problem is that AppVeyor runs the tests in serial. Can Travis run them in parallel like Circle? Or am I misunderstanding what Circle does?

I don’t think we should be merging after just one test passes on Windows. I have been bitten locally by differences between 3.5 and 2.7.

**Comment by jcfr** [comment]
> AppVeyor runs the tests in serial.

Paying for more worker node would help. 

$99/month -> 2 concurrent job
$300/month -> unlimited concurrent job


> don't think appveyor gives us that much freedom

You could conditionally run tox with a different env variable using logic like the following in each step:

```
- ps: |
      if ($env:appveyor_repo_tag -eq $true -and $env:appveyor_repo_tag_name –match "^[0-9]+(\.[0-9]+)*(\.post[0-9]+)?$") {
        Write-Host "build or test or deploy release .... "
      } elseif ($env:appveyor_repo_branch -eq "dev") {
         Write-Host "build or test or deploy dev ... "
      } else {
         Write-Host "build or test or deploy any other branch... "
      }
```

**Comment by t-b** [comment]
I think the main issue here is workflow-issue. In gitlab I can say "merge PR once CI was successful", so I don't need to wait until its done.

**Comment by mgrauer** [comment]
We could make changes here, but they do come with tradeoffs.

We could make it so the tests run post-merge or even nightly, then we will have to make sure somebody fixes any broken tests, and this might mean untangling multiple PRs that went in at the same time.

We could only run some tests before merge and then others after, but that becomes more difficult to maintain the CI, and also we have to decide what is appropriate to run before or after and why.

I believe (@dorukozturk please correct me here if I am wrong on this entire paragraph) we enforce the fact that any PR needs to be rebased/reforwarded on to the current master and have tests pass before merging, and we can change this so that if the tests had passed when the PR was originally submitted (based on master at that time and not as master now is) the PR could be merged (assuming there are not merge conflicts). This also has the problem in that merges to master could now be broken, and so someone will have to fix them after the fact.

Our goal with the CI work was to provide strictness to ensure software quality, especially since we are testing across multiple pythons and OSes, and this strictness comes with tradeoffs of taking more time to get through CI.

How do people want to proceed?

**Comment by t-b** [comment]
How about getting some numbers on how slow appveyor really is?

**Comment by oruebel** [comment]
> Our goal with the CI work was to provide strictness

Running the tests before merge is generally preferable. 

> How about getting some numbers on how slow appveyor really is?

Yes, it would be good to know how long it actually takes. Part of the problem I suspect is that we have a lot of PRs coming through right now due to the push to finalize the schema for 2.0 and release in general.

> Paying for more worker node would help.

If moving to another CI service or paying for a subscription is the solution then we can certainly look if we can find someone to sponsor the subscription.

**Comment by jcfr** [comment]
> some numbers on how slow appveyor really is?

Total time is ~25mins

![image](https://user-images.githubusercontent.com/219043/48232686-6c792d80-e380-11e8-8196-23d5f68936a1.png)


For comparison, the total time on CircleCI is ~6mins, and ~11mins on TravisCI  (see screenshot below)


> How do people want to proceed?

I suggest to remove Python 3.5 test on Appveyor. It would then be ~33% faster to get feedback with a wait of 15mins per PR



### CircleCI

![image](https://user-images.githubusercontent.com/219043/48232302-240d4000-e37f-11e8-9624-2b9567ee5f21.png)


### TravisCI

![image](https://user-images.githubusercontent.com/219043/48232356-4ef79400-e37f-11e8-9991-e747d098c61e.png)

**Comment by oruebel** [comment]
> I suggest to remove Python 3.5 test on Appveyor.

I think having Python 2.7 and 3.6 is essential. Removing Python 3.5 and only testing with 3.6 seems reasonable to me.

**Comment by mgrauer** [comment]
This is now done, removing Python 3.5 from appveyor, but people need to merge dev into their PR branches to get the speed up.

**Comment by bendichter** [comment]
Thanks guys!
